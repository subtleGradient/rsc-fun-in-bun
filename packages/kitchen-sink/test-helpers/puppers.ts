import packageJson from "#package"
import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser, ConsoleMessage, Page } from "puppeteer-core"

let browser: Browser

////////////////////////////////////////////////////////////////
// duck punching shennanigans

declare module "puppeteer-core" {
  interface Browser {
    [Symbol.asyncDispose](): Promise<void>
  }
  interface Page {
    [Symbol.asyncDispose](): Promise<void>
    pipeConsoleLogs: boolean
    $pipeConsoleLogs_isEnabled: boolean
  }
}

Browser.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (this.connected) await this.close()
  } catch (e) {
    console.warn("Browser.prototype[Symbol.asyncDispose]", e)
  }
}

Page.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (!this.isClosed()) await this.close({ runBeforeUnload: false })
  } catch (e) {
    console.warn("Page.prototype[Symbol.asyncDispose]", e)
  }
}

Object.defineProperties(Page.prototype, {
  pipeConsoleLogs_isEnabled: { configurable: true, enumerable: false, writable: true, value: false },
  pipeConsoleLogs: {
    configurable: true,
    enumerable: true,
    get(this: Page) { return this.$pipeConsoleLogs_isEnabled }, // prettier-ignore
    set(this: Page, isEnabled: boolean) {
      if ((this.$pipeConsoleLogs_isEnabled = isEnabled)) {
        this.on("console", async (msg: ConsoleMessage) => {
          const args = await Promise.all(msg.args().map(it => it.toString()))
          // @ts-ignore -- trust me bro ðŸ˜Ž
          console[msg.type()](...args)
        })
        this.on("error", console.error)
      } else {
        this.off("console")
        this.off("error")
      }
    },
  },
})

////////////////////////////////////////////////////////////////

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    dumpio: true,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
      `--user-data-dir=${process.env.TMPDIR ?? "/tmp/"}/puppers-${packageJson.name.replace(/[^a-z0-9_.$]/gi, "-")}`,
    ],
  })

  process.once("beforeExit", async () => await browser.close())
})

afterAll(() => browser.close())

export { browser, puppeteer }
