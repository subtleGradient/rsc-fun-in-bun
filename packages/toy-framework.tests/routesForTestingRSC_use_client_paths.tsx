export const routesForTestingRSC_use_client_paths = {
  rsc: "/rsc/test/use-client",
  render: "/rsc/test/use-client/render",
}
