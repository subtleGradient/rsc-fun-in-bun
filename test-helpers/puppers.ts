import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser } from "puppeteer-core"

let browser: Browser

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
    ],
  })
})

afterAll(() => browser.close())

// return Object.assign(browser, { [Symbol.dispose]: () => browser.close() })

export { browser, puppeteer }
