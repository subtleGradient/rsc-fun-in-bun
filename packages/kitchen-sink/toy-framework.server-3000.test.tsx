/// <reference lib="dom" />
import { describe, expect, it } from "bun:test"
import { routesForTestingRSC_use_client_paths } from "./routesForTestingRSC_use_client_paths.tsx"
import { browser } from "./test-helpers/puppers.ts"

const ROOT_DIRNAME = __dirname.split("/").slice(0, -2).join("/")
const RSC_TYPE = "text/x-component"

try {
  // trigger the tests to re-run when this file changes

  // BUG: Bun@1.1.4 does not support --conditions yet https://github.com/oven-sh/bun/issues/10036
  // So, I'm using a dynamic import below to avoid triggering the react-server error
  await import("./kitchen-sink.ts")
} catch (error) {}

if (!1!) {
  describe("the `await using` keyword", async () => {
    it("disposes of the page auto-magically", async () => {
      await using page = await browser.newPage()
      await page.goto("about:blank")
      // no need to call page.close() here! ðŸ¤¯
    })
  })
}

describe("toy-framework.server", () => {
  describe("root / route", () => {
    it("returns a response", async () => {
      const response = await fetch(new Request(`http://localhost:3000`))
      expect(response.status).toBe(200)
      expect((await response.text()).replaceAll("><", ">\n<")).toMatchSnapshot()
    })

    it("renders ClientComponent", async () => {
      console.time("browser.newPage")
      await using page = await browser.newPage()
      console.timeEnd("browser.newPage")
      page.goto(`http://localhost:3000`)
      await page.waitForSelector("#generated-by-client")
      expect(await page.$eval("#generated-by-client", el => el.textContent)).toBe("Hello from ClientComponent!")
    })
  })

  describe("/rsc/test-suspense", () => {
    it("returns RSC content", async () => {
      const response = await fetch(new Request(`http://localhost:3000/rsc/test-suspense`, { headers: { Accept: RSC_TYPE } }))
      expect(response.status).toBe(200)
      expect(await response.text()).toMatchSnapshot()
    })
    it(
      "updates asynchronously",
      async done => {
        const rscResponse = await fetch(
          new Request(`http://localhost:3000/rsc/test-suspense`, { headers: { Accept: RSC_TYPE } }),
        )
        // stream the responce and verify that it sends multiple chunks
        const decoder = new TextDecoder()
        const reader = rscResponse.body!.getReader()
        reader.closed.then(done)
        let count = 0
        while (true) {
          const { done, value } = await reader.read()
          if (done) break
          // expect(decoder.decode(value)).toMatchSnapshot()
          count++
        }
        expect(count).toBeGreaterThan(1)
      },
      { retry: 3 },
    )
  })

  describe("/rsc/test-suspense-render", () => {
    it("returns RSC content", async () => {
      const response = await fetch(new Request(`http://localhost:3000/rsc/test-suspense-render`))
      expect(response.status).toBe(200)
      expect(await response.text().then(text => text.replaceAll("><", ">\n<"))).toMatchSnapshot()
    })

    it("renders ClientComponent", async () => {
      console.time("browser.newPage")
      await using page = await browser.newPage()
      console.timeEnd("browser.newPage")

      console.time("page.goto")
      await page.goto(`http://localhost:3000/rsc/test-suspense-render`)
      console.timeEnd("page.goto")

      console.time("page.waitForSelector body")
      await page.waitForSelector("body")
      console.timeEnd("page.waitForSelector body")

      console.time("page.waitForSelector 0")
      await page.waitForSelector("#AsyncView0")
      console.timeEnd("page.waitForSelector 0")

      console.time("page.waitForSelector 10")
      await page.waitForSelector("#AsyncView10")
      console.timeEnd("page.waitForSelector 10")

      console.time("page.$eval")
      expect(await page.$eval("#AsyncView10", el => el.textContent)).toMatch(/slept for 10ms/)
      console.timeEnd("page.$eval")
    })
  })

  describe(routesForTestingRSC_use_client_paths.render, () => {
    const rsc_url = `http://localhost:3000${routesForTestingRSC_use_client_paths.rsc}`
    const render_url = `http://localhost:3000${routesForTestingRSC_use_client_paths.render}`

    it("exists", async () => {
      expect((await fetch(rsc_url)).status).toBe(200)
      expect((await fetch(render_url)).status).toBe(200)
    })

    describe("RSC payload", () => {
      it("mentions ClientComponent by name", async () => {
        const response = await fetch(rsc_url)
        const rscText = await response.text()
        expect(rscText).toMatch("ExampleClientComponent")
        expect(rscText).toMatch("examples/example0.client")
      })

      it("should not include absolute paths", async () => {
        const response = await fetch(rsc_url)
        const rscText = await response.text()

        try {
          // await import("#toy-framework/plugins/useClient_fromServer.plugin.ts")
        } catch (error) {}

        expect(rscText).not.toMatch(ROOT_DIRNAME) // no absolute paths
      })
    })

    it.todo("renders ClientComponent", async () => {
      console.time("browser.newPage")
      await using page = await browser.newPage()
      console.timeEnd("browser.newPage")

      console.time("page.goto")
      await page.goto(rsc_url)
      console.timeEnd("page.goto")

      console.time("page.waitForSelector body")
      await page.waitForSelector("body")
      console.timeEnd("page.waitForSelector body")

      console.time("page.waitForSelector #ClientComponent")
      await page.waitForSelector("#ClientComponent")
      console.timeEnd("page.waitForSelector #ClientComponent")

      console.time("page.$eval")
      expect(await page.$eval("#ClientComponent", el => el.textContent)).toBe("Hello from ClientComponent!")
      console.timeEnd("page.$eval")
    })
  })
})
