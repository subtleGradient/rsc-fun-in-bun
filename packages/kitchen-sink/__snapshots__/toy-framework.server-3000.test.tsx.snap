// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`toy-framework.server root / route returns a response 1`] = `
"<!DOCTYPE HTML>
<HTML lang=en>
<script type="importmap">{
  "imports": {
    "bun": "/!/externals:bun.mjs",
    "react": "/!/externals/react.mjs",
    "react-dom": "/!/externals/react-dom.mjs",
    "react-dom/client": "/!/externals/react-dom:client.mjs",
    "react-server-dom-webpack/client": "/!/externals/react-server-dom-webpack:client.mjs",
    "react/jsx-dev-runtime": "/!/externals/react:jsx-dev-runtime.mjs",
    "react/jsx-runtime": "/!/externals/react:jsx-runtime.mjs"
  }
}</script>
<head>
<title>Hello from /HTMLPageRootLayout.tsx</title>
<link rel="modulepreload" href="/!/polyfillsAndStuff.mjs"/>
<link rel="modulepreload" href="/!/externals/clientEntryPoint.externals.mjs"/>
<link rel="modulepreload" href="/!/externals/react.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react-server-dom-webpack:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-dev-runtime.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-runtime.mjs"/>
<link rel="modulepreload" href="/!/clientEntryPoint/clientEntryPoint.mjs"/>
<style>
          /* dark mode */
          @media (prefers-color-scheme: dark) {
            body {
              background-color: #333;
              color: #ddd;
            }
          }
        </style>
</head>
<body>
<h1>Hello from <!-- -->/HTMLPageRootLayout.tsx</h1>
<div id="root">
<div data-id="NOT-generated-by-client">pre-rendered by server in <code>/HTMLPageRootLayout.tsx</code>
</div>
</div>
<script src="/!/polyfillsAndStuff.mjs">
</script>
<script type="module" src="/!/externals/clientEntryPoint.externals.mjs">
</script>
<script type="module" src="/!/clientEntryPoint/clientEntryPoint.mjs">
</script>
</body>
</HTML>"
`;
